framework/filters/AccessRule.php

```diff
@@ -161,8 +191,13 @@ protected function matchRole($user)
                 if (!$user->getIsGuest()) {
                     return true;
                 }
-            } elseif ($user->can($role)) {
-                return true;
+            } else {
+                if (!isset($roleParams)) {
```

Comment:
```php
            if ($role === '?') {
                if ($user->getIsGuest()) {
                    return true; // this line always return true even if the creator ip are not the same.
                }
            } elseif ($role === '@') {
                if (!$user->getIsGuest()) {
                    return true; // this line always return true even if the creator id are not the same.
                }
            } else {
                if (!isset($roleParams)) {
                    $roleParams = $this->roleParams instanceof Closure ? call_user_func($this->roleParams, $this) : $this->roleParams;
                }
                if ($user->can($role, $roleParams)) {
                    return true;
                }
            }
```
i checked the DbManager and seems it does the same

https://github.com/yiisoft/yii2/blob/master/framework/rbac/DbManager.php#L125

unauth users or general roles get rejected without using the `$params` property.